### PR TITLE
[Api] Nullable

### PR DESCRIPTION
**Pull request by ysolomchenko**
Towards https://github.com/open-telemetry/opentelemetry-dotnet/issues/3958

## Changes

OpenTelemety.Api mark as nullable

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Review by Kielek — APPROVED** [review]
LGTM with fixes.
While reviewing, please focus on Baggage and RuntimeContext related things.

File: src/OpenTelemetry/Metrics/PullMetricScope.cs [hunk 1]
```diff
@@ -7,7 +7,7 @@ namespace OpenTelemetry.Metrics;
 
 internal sealed class PullMetricScope : IDisposable
 {
-    private static readonly RuntimeContextSlot<bool> Slot = RuntimeContext.RegisterSlot<bool>("otel.pull_metric");
+    private static readonly RuntimeContextSlot<bool> Slot = RuntimeContext.RegisterSlot<bool>("otel.pull_metric")!;
```
Comment (TimothyMothra) [inline src/OpenTelemetry/Metrics/PullMetricScope.cs:10]: I think this one needs a closer look.
I'm not familiar with the RuntimeContext api. If the RegisterSlot method can return a nullable, that might suggest something failed.

File: src/Shared/Guard.cs [hunk 2]
```diff
@@ -19,12 +19,12 @@ namespace System.Runtime.CompilerServices
     [AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = false)]
     internal sealed class CallerArgumentExpressionAttribute : Attribute
     {
-        public CallerArgumentExpressionAttribute(string parameterName)
+        public CallerArgumentExpressionAttribute(string? parameterName)
```
Comment (TimothyMothra) [inline src/Shared/Guard.cs:22]: Should doublecheck this. The original class doesn't have nullable. ([link](https://github.com/dotnet/runtime/blob/b14e2f5ef3008ab62bdbe8044f00fe948e20595e/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CallerArgumentExpressionAttribute.cs#L12))

My understanding is that this should never be null.

Comment (Kielek) [reply src/Shared/Guard.cs:22]: See also https://github.com/dotnet/runtime/blob/v9.0.0-rc.1.24431.7/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/CallerArgumentExpressionAttributeTests.cs (the same tests we have copied to this repository).

I am not against dropping support for null, but it is kind of break for what we have in dotnet repository.

Whatever we decide, the impact should be low. It is extremely internal class. Any recommendation based on this?

File: src/OpenTelemetry/Metrics/PullMetricScope.cs (hunk 1 above)
Comment (Kielek) [reply src/OpenTelemetry/Metrics/PullMetricScope.cs]: Fixed in fe4d9bf9e1740bdd673aa0ae68281d1678f39bdb

I double checked. There is no possibility to return null (at least as long someone does no play with reflection with `contextSlotType` field in `RuntimeContext` class.

File: src/Shared/Guard.cs (hunk 2 above)
Comment (TimothyMothra) [reply src/Shared/Guard.cs:22]: Thanks. I think this changes is good.

**Review by TimothyMothra — APPROVED** [review]
these changes LGTM.

One thing to note, I don't know the `Baggage` or `RuntimeContext` classes well enough to say if these values SHOULD be nullable.

**Pull request merged**